src/Protocol/ForgeHandshake.cpp

```diff
@@ -0,0 +1,366 @@
+{
+	LOGD("Received ModList");
+
+	AStringMap ClientMods = ParseModList(a_Data + 1, a_Size - 1);
+	AString ClientModsString;
+	for (auto & item: ClientMods)
+	{
+		ClientModsString.append(item.first);
+		ClientModsString.append("@");
+		ClientModsString.append(item.second);
+		ClientModsString.append(", ");
+	}
+
+	LOG("Client connected with " SIZE_T_FMT " mods: %s", ClientMods.size(), ClientModsString.c_str());
+
+	m_Client->m_ForgeMods = ClientMods;
+
+	// Let the plugins know about this event, they may refuse the player:
+	if (cRoot::Get()->GetPluginManager()->CallHookLoginForge(*a_Client, ClientMods))
+	{
+		LOG("Modded client refused by plugin");
+		SetError();
+		return;
+	}
+
+	// Send server ModList
+
+	// TODO: max size?
+	cByteBuffer Buf(1024);
+
+	// Send server-side Forge mods registered by plugins
+	auto &ServerMods = m_Client->GetForgeMods();
+
+	size_t ModCount = ServerMods.size();
+
+	Buf.WriteBEInt8(Discriminator::ModList);
+	Buf.WriteVarInt32(static_cast<UInt32>(ModCount));
+	for (auto & item: ServerMods)
+	{
+		const AString & name = item.first;
+		const AString & version = item.second;
+		Buf.WriteVarUTF8String(name);
+		Buf.WriteVarUTF8String(version);
+	}
+	AString ServerModList;
+	Buf.ReadAll(ServerModList);
+
+	m_Client->SendPluginMessage("FML|HS", ServerModList);
+}
+
+
+
+
+void cForgeHandshake::HandleHandshakeAck(cClientHandle * a_Client, const char * a_Data, size_t a_Size)
+{
+	if (a_Size != 2)
+	{
+		LOG("Unexpected HandshakeAck packet length: " SIZE_T_FMT "", a_Size);
+		SetError();
+		return;
+	}
+
+	int Phase = a_Data[1];
+	LOGD("Received client HandshakeAck with phase=%d", Phase);
+
+	switch (Phase)
+	{
+		case ClientPhase::WAITINGSERVERDATA:
+		{
+			cByteBuffer Buf(1024);
+			Buf.WriteBEInt8(Discriminator::RegistryData);
+
+			// TODO: send real registry data
+			bool HasMore = false;
+			AString RegistryName = "potions";
+			UInt32 NumIDs = 0;
+			UInt32 NumSubstitutions = 0;
+			UInt32 NumDummies = 0;
+
+			Buf.WriteBool(HasMore);
+			Buf.WriteVarUTF8String(RegistryName);
+			Buf.WriteVarInt32(NumIDs);
+			Buf.WriteVarInt32(NumSubstitutions);
+			Buf.WriteVarInt32(NumDummies);
+
+			AString RegistryData;
+			Buf.ReadAll(RegistryData);
+			m_Client->SendPluginMessage("FML|HS", RegistryData);
+			break;
+		}
+
+		case ClientPhase::WAITINGSERVERCOMPLETE:
+		{
+			LOGD("Client finished receiving registry data; acknowledging");
+
+			AString Ack;
+			Ack.push_back(Discriminator::HandshakeAck);
+			Ack.push_back(ServerPhase::WAITINGCACK);
+			m_Client->SendPluginMessage("FML|HS", Ack);
+			break;
+		}
+
+		case ClientPhase::PENDINGCOMPLETE:
+		{
+			LOGD("Client is pending completion; sending complete ack");
+
+			AString Ack;
+			Ack.push_back(Discriminator::HandshakeAck);
+			Ack.push_back(ServerPhase::COMPLETE);
+			m_Client->SendPluginMessage("FML|HS", Ack);
+
+			// Now finish logging in
+			m_Client->FinishAuthenticate(m_Name, m_UUID, m_Properties);
+			break;
+		}
+	}
```

Comment:
This revealed the ClientPhase::COMPLETE phase was received but not handled; added default and moved FinishAuthenticate in case ClientPhase::COMPLETE